ankimorphs/jit_highlight_morphs.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+import re
+
+from anki.template import TemplateRenderContext
+
+from .ankimorphs_config import (
+    AnkiMorphsConfig,
+    AnkiMorphsConfigFilter,
+    get_matching_filter,
+)
+from .ankimorphs_db import AnkiMorphsDB
+from .ankimorphs_globals import EXTRA_FIELD_HIGHLIGHTED
+from .morpheme import Morpheme
+from .morphemizers import spacy_wrapper
+from .morphemizers.morphemizer import (
+    Morphemizer,
+    SpacyMorphemizer,
+    get_morphemizer_by_description,
+)
+from .text_highlighting import get_highlighted_text
+from .text_preprocessing import (
+    get_processed_morphemizer_morphs,
+    get_processed_spacy_morphs,
+)
+
+
+def am_highlight_morphs(
+    field_text: str,
+    field_name: str,
+    filter_name: str,
+    context: TemplateRenderContext,
+) -> str:
+    """Use morph learning progress to decorate the morphemes in the supplied text.
+    Adds css classes to the output that can be styled in the card."""
+
+    if filter_name != "am-highlight" or field_name == EXTRA_FIELD_HIGHLIGHTED:
+        return field_text
+
+    am_config_filter: AnkiMorphsConfigFilter | None = get_matching_filter(
+        context.note()
+    )
+
+    if am_config_filter is None:
+        return field_text
+
+    morphemizer: Morphemizer | None = get_morphemizer_by_description(
+        am_config_filter.morphemizer_description
+    )
+
+    if not morphemizer:
+        return field_text
+
+    am_config = AnkiMorphsConfig()
+
+    card_morphs: list[Morpheme] = get_morphemes(morphemizer, am_config, field_text)
+
+    if not card_morphs:
+        return field_text
+
+    return get_highlighted_text(am_config, card_morphs, field_text)
```

Comment:
I prefer qualified naming whenever it's not super obnoxiously long, it really improves clarify.

```python
import .text_highlighting
...
return text_highlighting.get_highlighted_text(am_config, card_morphs, field_text)
```

Now that I think about it, it might be worth considered moving all this code to `text_highlighting.py`, provided it doesn't lead to any cyclical imports... what do you think?

Reply:
RE: fully qualified imports, I did notice that at one point our styles differed there, shame on me for not taking the 30 seconds to adopt your style in your codebase; will do.

RE: moving into `text_highlighting`. I don't know how a feel about that. I like that the template filter(s) stand alone, they are a new kind of entity, and could warrant separation. I think it COULD go into `text_highlighting` I just don't think it's a slam dunk.

I think `dehtml` makes sense to move there, it's a pure function, and just manipulates text, so..... it's a good candidate.

I think `update_morphs` should stay with the filter itself (imo). it's a little tacky in that it ONLY fetches a tiny bit of metadata required by the text formatter, so in that manner it does not deserve "promotion" into the main libraries.

I think `get_morphemes` should probably move to the morphemes file. (side note, I ALMOST did a refactor to extract this code from yours where you perform this operation en masse durning recalls, so that we could keep the code DRY, but remembered what you said about small updates, so we have this non-dry code)

what are your thoughts?

Reply:
I did everything else in the review besides these moves.

I wanted to give you a moment to think about my comments, then I'll go with whatever your decision is.

Reply:
> I think it COULD go into `text_highlighting` I just don't think it's a slam dunk.

Yeah, it's kind of an awkward in-between. We can leave it as is for now.

> I think `get_morphemes` should probably move to the morphemes file. (side note, I ALMOST did a refactor to extract this code from yours where you perform this operation en masse durning recalls

Currently it's a good duplication imo, I don't want to do anything that might slow down recalc.